### ISSUE TITLE
Investigate `StorageError: invalid object, Code: 4, Key`

### ISSUE THREAD

**Issue by alexec** [argoproj/argo#3158]
This error sporadically fails CI builds. I've not seen this before and expect it due to hydrator changes:

https://circleci.com/api/v1.1/project/github/argoproj/argo/26101/output/111/0?file=true&allocation-id=5ed6ea8b5d1b4f7dcaa57c85-0-build%2F6A140EDD

```
controller | time="2020-06-03T00:18:00Z" level=warning msg="Workflow 'argo/test-7tt6h' no longer exists"
 controller | time="2020-06-03T00:18:00Z" level=warning msg="Workflow 'argo/test-ptvtg' no longer exists"
 controller | 2020/06/03 00:18:00 
 controller | 	Session ID:     00001
 controller | 	Query:          DELETE FROM "argo_workflows" WHERE ("clustername" = $1 AND "uid" = $2 AND "version" <> $3 AND updatedat < current_timestamp - interval '30' second)
 controller | 	Arguments:      []interface {}{"default", "047d0a9f-507e-4fc7-b6d5-ee066331e361", "fnv:2551433244"}
 controller | 	Rows affected:  0
 controller | 	Time taken:     0.09467s
 controller | 	Context:        context.Background
 controller | time="2020-06-03T00:18:00Z" level=debug msg="Deleted offloaded nodes" rowsAffected=0 uid=047d0a9f-507e-4fc7-b6d5-ee066331e361 version="fnv:2551433244"
 controller | time="2020-06-03T00:18:00Z" level=warning msg="Error updating workflow: Operation cannot be fulfilled on workflows.argoproj.io \"test-7tt6h\": StorageError: invalid object, Code: 4, Key: /registry/argoproj.io/workflows/argo/test-7tt6h, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 047d0a9f-507e-4fc7-b6d5-ee066331e361, UID in object meta:  Conflict" namespace=argo workflow=test-7tt6h
 controller | time="2020-06-03T00:18:00Z" level=info msg="Re-applying updates on latest version and retrying update" namespace=argo workflow=test-7tt6h
 controller | time="2020-06-03T00:18:00Z" level=info msg="Failed to re-apply update: workflows.argoproj.io \"test-7tt6h\" not found" namespace=argo workflow=test-7tt6h
 controller | time="2020-06-03T00:18:00Z" level=debug msg="Getting offloaded nodes" uid=d70a69ee-bb04-4741-958d-77da1695da26 version="fnv:2931735855"
 controller | 2020/06/03 00:18:00 
 controller | 	Session ID:     00001
 controller | 	Query:          SELECT * FROM "argo_workflows" WHERE ("clustername" = $1 AND "uid" = $2 AND "version" = $3)
 controller | 	Arguments:      []interface {}{"default", "d70a69ee-bb04-4741-958d-77da1695da26", "fnv:2931735855"}
 controller | 	Time taken:     0.00293s
 controller | 	Context:        context.Background
 controller | time="2020-06-03T00:18:00Z" level=error msg="hydration failed: upper: no more rows in this result set" namespace=argo workflow=test-8t2mp
 controller | time="2020-06-03T00:18:00Z" level=info msg="Updated phase Running -> Error" namespace=argo workflow=test-8t2mp
 controller | time="2020-06-03T00:18:00Z" level=info msg="Updated message  -> upper: no more rows in this result set" namespace=argo workflow=test-8t2mp
 controller | time="2020-06-03T00:18:00Z" level=info msg="Marking workflow completed" namespace=argo workflow=test-8t2mp
 controller | E0603 00:18:00.240670   19522 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
 controller | goroutine 315 [running]:
 controller | k8s.io/apimachinery/pkg/util/runtime.logPanic(0x172fa60, 0x285d290)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/runtime/runtime.go:74 +0xa3
 controller | k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/runtime/runtime.go:48 +0x82
 controller | panic(0x172fa60, 0x285d290)
 controller | 	/home/circleci/sdk/go1.13.4/src/runtime/panic.go:679 +0x1b2
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).deletePDBResource(0xc0006899a0, 0x1941144, 0x9)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:2619 +0x41
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).markWorkflowPhase(0xc0006899a0, 0x193dcb0, 0x5, 0x1, 0xc000cdb958, 0x1, 0x1)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:1619 +0x639
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).markWorkflowError(0xc0006899a0, 0x1b96c00, 0xc0002c0c40, 0x1)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:1657 +0x99
 controller | github.com/argoproj/argo/workflow/controller.(*WorkflowController).processNextItem(0xc000140800, 0x0)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:432 +0xe75
 controller | github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker(0xc000140800)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:365 +0x2b
 controller | k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0009e0210)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:152 +0x5e
 controller | k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009e0210, 0x3b9aca00, 0x0, 0x1, 0xc00068d020)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:153 +0xf8
 controller | k8s.io/apimachinery/pkg/util/wait.Until(0xc0009e0210, 0x3b9aca00, 0xc00068d020)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:88 +0x4d
 controller | created by github.com/argoproj/argo/workflow/controller.(*WorkflowController).Run
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:178 +0x9cd
 controller | panic: runtime error: invalid memory address or nil pointer dereference [recovered]
 controller | 	panic: runtime error: invalid memory address or nil pointer dereference
 controller | [signal SIGSEGV: segmentation violation code=0x1 addr=0x1a8 pc=0x1595251]
 controller | goroutine 315 [running]:
 controller | k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/runtime/runtime.go:55 +0x105
 controller | panic(0x172fa60, 0x285d290)
 controller | 	/home/circleci/sdk/go1.13.4/src/runtime/panic.go:679 +0x1b2
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).deletePDBResource(0xc0006899a0, 0x1941144, 0x9)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:2619 +0x41
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).markWorkflowPhase(0xc0006899a0, 0x193dcb0, 0x5, 0x1, 0xc000cdb958, 0x1, 0x1)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:1619 +0x639
 controller | github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).markWorkflowError(0xc0006899a0, 0x1b96c00, 0xc0002c0c40, 0x1)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/operator.go:1657 +0x99
 controller | github.com/argoproj/argo/workflow/controller.(*WorkflowController).processNextItem(0xc000140800, 0x0)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:432 +0xe75
 controller | github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker(0xc000140800)
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:365 +0x2b
 controller | k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0009e0210)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:152 +0x5e
 controller | k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009e0210, 0x3b9aca00, 0x0, 0x1, 0xc00068d020)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:153 +0xf8
 controller | k8s.io/apimachinery/pkg/util/wait.Until(0xc0009e0210, 0x3b9aca00, 0xc00068d020)
 controller | 	/home/circleci/.go_workspace/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:88 +0x4d
 controller | created by github.com/argoproj/argo/workflow/controller.(*WorkflowController).Run
 controller | 	/home/circleci/.go_workspace/src/github.com/argoproj/argo/workflow/controller/controller.go:178 +0x9cd
 controller | Terminating controller
```

See #3118 
See #3123

**Comment by alexec** [comment]
It appears that the workflow was deleted when we attempted to call `woc.persistUpdates`, but `Workflows.Update()` returned `Conflict`, not NotFound (as you'd expect) and so we attempt to re-apply updates. At this point it appears we've lost all the data we need to correctly update the workflow.

**Comment by alexec** [comment]
This was a cron workflow because `"Workflow '%s' no longer exists"` is only logged in `cron/controller.go`.

**Comment by alexec** [comment]
we're attempting to operate on a deleted workflow, fail and panic - we should be able to deal with this without panicing

**Comment by alexec** [comment]
Duplicate of #3121

**Comment by sarabala1979** [comment]
let me take a look

**Comment by sarabala1979** [comment]
Fixed in #3160
